keps/sig-cluster-lifecycle/wgs/1815-componentconfig-objectmeta/README.md

```diff
@@ -0,0 +1,577 @@
+possible to collect this information before implementation begins in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+The proposal is for adding standard ObjectMeta to the ComponentConfig API's to follow the standard pattern for other Kubernetes API's. This will allow for better interpolation with standard tooling like kustomize for managing config templating and substitution.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The initial implementation of the ComponentConfig API's didn't include OjectMeta fields. The Kubernetes project tooling relies on using Kustomize with manually added ObjectMeta fields in the ComponentConfig manifests. This functionality was broken as strict decoders were implemented for ComponentConfig API's and forced a leninent path to be added for the current ComponentConfig API versions. Adding ObjectMeta will allow for valid usage with Kustomize and strict decoders for current and future ComponentConfig API's.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Add a standard set of ObjectMeta fields (name, labels, and annotations) to the ComponentConfig API's.
+  - kubelet
+  - kube-proxy
+  - kube-scheduler
+  - kube-controller-manager
+  - kube-apiserver
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Graduate the API versions
+  - We will implement ObjectMeta within the current version of each ComponetConfig API
+- Implement every possible ObjectMeta field for ComponentConfig API's
+  - As these configs are singular per component, per cluster, they won't have or need the full suite of ObjectMeta similar to core Kubernetes API's (ie. Pods, Servcies, etc.)
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- for component in [kubelet kubeproxy kubecontrollermanager kubeapiserver kubescheduler]
+  - Add the following ObjectMeta fields
+    - name
+    - labels
+    - annotations
```

Comment:
This is a great start.

Let's add some examples on what this would look like in terms of additions to the Go structs, and in-practice in yaml form.

Let's also consider where to put the definition of the ComponentConfig subset of ObjectMeta (pick a reasonable location in the k8s.io/component-base repo).

Reply:
it's unclear to me what these fields *mean* in the config... I wouldn't expect to add them just to support putting dummy values in them

Reply:
I added additional info on the struct additions and practical examples in to the design section.

I have some questions around proper place in the component-base repo, will followup with @mtaufen on that. We can discuss viability of fields beyond `metadata.name` as well.

Reply:
I think it might be fine to start with `metadata.name`.

Maybe @rosti or @neolit123 can point to examples where users are currently adding meta fields for use with kustomize; are they just adding names or are they doing more than that?

Reply:
AFAIK they are adding names, just to conform kustomize.
kustomize requires metadata.name to be able to patch any Kind.

this is more of a question whether we should standardize ComponentConfigs to always have this field to conform kustomize, as supposedly kustomize is a prime citizen of the eco.

we still track adding ObjectMeta to the kubeadm CC here https://github.com/kubernetes/kubeadm/issues/1727
that said, projects like kubeadm and kind moved away from kustomize and implemented raw patches for their internal "customization" mechanics.

Reply:
Removed `labels` and `annotations` to leave `name` only.